core/domain/question_domain.py

```diff
@@ -122,73 +128,71 @@ def to_dict(self):
-            )
+                'Expected version to be an integer, received %s' %
+                self.version)
 
-        if dest_is_specified:
+        if not any([self.language_code == lc['code']
+                    for lc in constants.ALL_LANGUAGE_CODES]):
             raise utils.ValidationError(
-                'Expected all answer groups to have destination as None.'
-            )
+                'Invalid language code: %s' % self.language_code)
 
-        if (len(interaction['hints']) == 0) or (
-                interaction['solution'] is None):
-            raise utils.ValidationError(
-                'Expected the question to have at least one hint and a ' +
-                'solution.'
-            )
+        if strict:
+            at_least_one_correct_answer = False
+            dest_is_specified = False
+            interaction = self.question_state_data['interaction']
+            for answer_group in interaction['answer_groups']:
+                if answer_group['labelled_as_correct']:
+                    at_least_one_correct_answer = True
+                if answer_group['dest'] is not None:
+                    dest_is_specified = True
+
+            if interaction['default_outcome']['labelled_as_correct']:
+                at_least_one_correct_answer = True
 
-        question_data = exp_domain.State.from_dict(self.question_data)
-        question_data.validate({}, True)
+            if interaction['default_outcome']['dest'] is not None:
+                dest_is_specified = True
 
-        if not isinstance(self.question_data_schema_version, int):
-            raise utils.ValidationError(
-                'Expected question_data_schema_version to be a integer,' +
-                'received %s' % self.question_data_schema_version)
+            if not at_least_one_correct_answer:
+                raise utils.ValidationError(
+                    'Expected at least one answer group to have a correct ' +
+                    'answer.'
+                )
 
-        if not isinstance(self.language_code, basestring):
-            raise utils.ValidationError(
-                'Expected language_code to be a string, received %s' %
-                self.language_code)
+            if dest_is_specified:
+                raise utils.ValidationError(
+                    'Expected all answer groups to have destination as None.'
+                )
 
-        if not any([self.language_code == lc['code']
-                    for lc in constants.ALL_LANGUAGE_CODES]):
-            raise utils.ValidationError(
-                'Invalid language code: %s' % self.language_code)
+            if (len(interaction['hints']) == 0) or (
+                    interaction['solution'] is None):
+                raise utils.ValidationError(
+                    'Expected the question to have at least one hint and a ' +
+                    'solution.'
+                )
+
+            question_state_data = exp_domain.State.from_dict(
```

Comment:
I thought self.question_state_data was already a State object? If that's the case, why are we calling from_dict() on it?

I think you have some issues with types here... or your docstrings/varnames are wrong, or something. Please check this very carefully. In particular your docstrings must be correct.